File: androidx/src/androidTest/java/me/panpf/androidx/test/app/FragmentxTest.java
/*
 * Copyright (C) 2018 Peng fei Pan <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.panpf.androidx.test.app;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.test.rule.ActivityTestRule;
import androidx.test.runner.AndroidJUnit4;
import androidx.viewpager.widget.ViewPager;

import org.junit.Assert;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import me.panpf.androidx.Androidx;
import me.panpf.androidx.app.Argsx;
import me.panpf.androidx.app.Fragmentx;
import me.panpf.androidx.os.BundleBuilder;
import me.panpf.androidx.test.R;
import me.panpf.androidx.util.ResultRunnable;
import me.panpf.javax.lang.Stringx;
import me.panpf.javax.util.Premisex;

@RunWith(AndroidJUnit4.class)
public class FragmentxTest {

    @NonNull
    private final ActivityTestRule<TestActivity> activityTestRule = new ActivityTestRule<>(TestActivity.class);
    @NonNull
    private final ActivityTestRule<TestFindUserVisibleChildActivity> findUserVisibleChildActivityRule = new ActivityTestRule<>(TestFindUserVisibleChildActivity.class);

    @Rule
    @NonNull
    public ActivityTestRule<TestActivity> getActivityTestRule() {
        return activityTestRule;
    }

    @Rule
    @NonNull
    public ActivityTestRule<TestFindUserVisibleChildActivity> getFindUserVisibleChildActivityRule() {
        return findUserVisibleChildActivityRule;
    }

    @Test
    public void testDestroyed() throws InterruptedException {
        android.app.Fragment originFragment = activityTestRule.getActivity().getOriginFragment();
        Fragment supportFragment = activityTestRule.getActivity().getSupportFragment();

        Assert.assertFalse(Fragmentx.isDestroyedCompat(originFragment));
        Assert.assertFalse(Fragmentx.isDestroyedCompat(supportFragment));

        activityTestRule.finishActivity();
        Thread.sleep(2000);

        Assert.assertTrue(Fragmentx.isDestroyedCompat(originFragment));
        Assert.assertTrue(Fragmentx.isDestroyedCompat(supportFragment));
    }

    @Test
    public void testGetImplWithParent() {
        TestImplOriginFragment originFragment = (TestImplOriginFragment) activityTestRule.getActivity().getOriginFragment();
        TestImplSupportFragment supportFragment = (TestImplSupportFragment) activityTestRule.getActivity().getSupportFragment();

        Assert.assertEquals(Premisex.requireNotNull(Fragmentx.getImplFromParent(originFragment, ImplTestInterface.class)).getClass(), TestImplOriginFragment.class);
        Assert.assertEquals(Premisex.requireNotNull(Fragmentx.getImplFromParent(supportFragment, ImplTestInterface.class)).getClass(), TestImplSupportFragment.class);

        if (Androidx.isAtLeast17()) {
            Assert.assertEquals(Premisex.requireNotNull(Fragmentx.getImplFromParent(originFragment.getChildFragment(), ImplTestInterface.class)).getClass(), TestImplOriginFragment.class);
        }
        Assert.assertEquals(Premisex.requireNotNull(Fragmentx.getImplFromParent(supportFragment.getChildFragment(), ImplTestInterface.class)).getClass(), TestImplSupportFragment.class);

        Androidx.waitRunInUI(new Runnable() {
            @Override
            public void run() {
                activityTestRule.getActivity().convertChildFragment();
            }
        });

        TestImplOriginFragment2 originFragment2 = Androidx.waitRunInUIResult(new ResultRunnable<TestImplOriginFragment2>() {
            @NonNull
            @Override
            public TestImplOriginFragment2 run() {
                return (TestImplOriginFragment2) activityTestRule.getActivity().getOriginFragment();
            }
        });
        TestImplSupportFragment2 supportFragment2 = Androidx.waitRunInUIResult(new ResultRunnable<TestImplSupportFragment2>() {
            @NonNull
            @Override
            public TestImplSupportFragment2 run() {
                return (TestImplSupportFragment2) activityTestRule.getActivity().getSupportFragment();
            }
        });

        Assert.assertEquals(Premisex.requireNotNull(Fragmentx.getImplFromParent(originFragment2, ImplTestInterface.class)).getClass(), TestActivity.class);
        Assert.assertEquals(Premisex.requireNotNull(Fragmentx.getImplFromParent(supportFragment2, ImplTestInterface.class)).getClass(), TestActivity.class);

        Assert.assertNull(Fragmentx.getImplFromParent(new TestImplOriginFragment2(), ImplTestInterface.class));
        Assert.assertNull(Fragmentx.getImplFromParent(new TestImplSupportFragment2(), ImplTestInterface.class));
    }

    @Test
    public void testInstantiate() {
        android.app.Fragment originFragment = Fragmentx.instantiateOrigin(android.app.Fragment.class, new BundleBuilder().putString("key", "testInstantiate").build());
        Assert.assertEquals(android.app.Fragment.class.getName(), originFragment.getClass().getName());
        Assert.assertEquals("testInstantiate", Argsx.readStringArg(originFragment, "key"));

        Fragment supportFragment = Fragmentx.instantiate(Fragment.class, new BundleBuilder().putString("key", "testInstantiate").build());
        Assert.assertEquals(Fragment.class.getName(), supportFragment.getClass().getName());
        Assert.assertEquals("testInstantiate", Argsx.readStringArg(supportFragment, "key"));

        android.app.Fragment originFragment2 = Fragmentx.instantiateOrigin(android.app.Fragment.class);
        Assert.assertEquals(android.app.Fragment.class.getName(), originFragment2.getClass().getName());

        Fragment supportFragment2 = Fragmentx.instantiate(Fragment.class);
        Assert.assertEquals(Fragment.class.getName(), supportFragment2.getClass().getName());
    }

    @Test
    public void testFindUserVisibleChildFragment() {
        TestFindUserVisibleChildActivity activity = findUserVisibleChildActivityRule.getActivity();
        // 定义多少个 ActivityTestRule 测试方法执行的时候就会启动多少个 ActivityTestRule 为了让 findUserVisibleChildActivityRule 处于 resumed 状态
        activityTestRule.finishActivity();
        // 此测试要求手机处于屏幕解锁状态

        Fragment fragmentFromActivity = Premisex.requireNotNull(Fragmentx.findUserVisibleChildFragment(activity));
        Assert.assertEquals(TestFindUserVisibleChildFragment.class.getName(), fragmentFromActivity.getClass().getName());

        Fragment fragmentFromList = Premisex.requireNotNull(Fragmentx.findUserVisibleChildFragment(activity.getSupportFragmentManager().getFragments()));
        Assert.assertEquals(TestFindUserVisibleChildFragment.class.getName(), fragmentFromList.getClass().getName());

        TestFindUserVisibleChildFragment fragmentFromActivity2 = (TestFindUserVisibleChildFragment) fragmentFromActivity;
        Fragment fragmentFromChildFragment = Premisex.requireNotNull(Fragmentx.findUserVisibleChildFragment(fragmentFromActivity2));
        Assert.assertTrue(fragmentFromChildFragment.getTag(), Stringx.orEmpty(fragmentFromChildFragment.getTag()).startsWith("android:switcher") && Stringx.orEmpty(fragmentFromChildFragment.getTag()).endsWith(":3"));
    }

    @Test
    public void testFindFragmentByViewPagerCurrentItem() {
        TestFindUserVisibleChildActivity activity = findUserVisibleChildActivityRule.getActivity();
        // 定义多少个 ActivityTestRule 测试方法执行的时候就会启动多少个 ActivityTestRule 为了让 findUserVisibleChildActivityRule 处于 resumed 状态
        activityTestRule.finishActivity();
        // 此测试要求手机处于屏幕解锁状态

        Fragment fragmentFromActivity = Premisex.requireNotNull(Fragmentx.findFragmentByViewPagerCurrentItem(activity, 2));
        Assert.assertEquals(TestFindUserVisibleChildFragment.class.getName(), fragmentFromActivity.getClass().getName());
        Assert.assertTrue(fragmentFromActivity.getTag(), Stringx.orEmpty(fragmentFromActivity.getTag()).startsWith("android:switcher") && Stringx.orEmpty(fragmentFromActivity.getTag()).endsWith(":2"));

        Fragment fragmentFromList = Premisex.requireNotNull(Fragmentx.findFragmentByViewPagerCurrentItem(activity.getSupportFragmentManager().getFragments(), 2));
        Assert.assertEquals(TestFindUserVisibleChildFragment.class.getName(), fragmentFromList.getClass().getName());
        Assert.assertTrue(fragmentFromList.getTag(), Stringx.orEmpty(fragmentFromList.getTag()).startsWith("android:switcher") && Stringx.orEmpty(fragmentFromList.getTag()).endsWith(":2"));

        TestFindUserVisibleChildFragment fragmentFromActivity2 = (TestFindUserVisibleChildFragment) fragmentFromActivity;
        Fragment fragmentFromChildFragment = Premisex.requireNotNull(Fragmentx.findFragmentByViewPagerCurrentItem(fragmentFromActivity2, 3));
        Assert.assertTrue(fragmentFromChildFragment.getTag(), Stringx.orEmpty(fragmentFromChildFragment.getTag()).startsWith("android:switcher") && Stringx.orEmpty(fragmentFromChildFragment.getTag()).endsWith(":3"));
    }

    public interface ImplTestInterface {
    }

    public static class TestActivity extends FragmentActivity implements ImplTestInterface {

        @Override
        protected void onCreate(@Nullable Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            setContentView(R.layout.at_multi_frame);

            getFragmentManager().beginTransaction()
                    .replace(R.id.multiFrameAt_frame1, new TestImplOriginFragment())
                    .commit();

            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.multiFrameAt_frame2, new TestImplSupportFragment())
                    .commit();
        }

        @NonNull
        public android.app.Fragment getOriginFragment() {
            return getFragmentManager().findFragmentById(R.id.multiFrameAt_frame1);
        }

        @NonNull
        public Fragment getSupportFragment() {
            return Premisex.requireNotNull(getSupportFragmentManager().findFragmentById(R.id.multiFrameAt_frame2));
        }

        public final void convertChildFragment() {
            getFragmentManager().beginTransaction()
                    .replace(R.id.multiFrameAt_frame1, new FragmentxTest.TestImplOriginFragment2())
                    .commit();
            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.multiFrameAt_frame2, new FragmentxTest.TestImplSupportFragment2())
                    .commit();
        }
    }

    public static final class TestImplOriginFragment extends android.app.Fragment implements FragmentxTest.ImplTestInterface {

        @Nullable
        public View onCreateView(@Nullable LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            return inflater != null ? inflater.inflate(R.layout.at_test, container, false) : null;
        }

        public void onViewCreated(@Nullable View view, @Nullable Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            if (Androidx.isAtLeast17()) {
                this.getChildFragmentManager().beginTransaction().replace(R.id.testAt_frame, new FragmentxTest.TestImplOriginFragment2())
                        .commit();
            }
        }

        @NonNull
        public final android.app.Fragment getChildFragment() {
            return Premisex.requireNotNull(getChildFragmentManager().findFragmentById(R.id.testAt_frame));
        }
    }

    public static final class TestImplOriginFragment2 extends android.app.Fragment {
    }

    public static final class TestImplSupportFragment extends Fragment implements FragmentxTest.ImplTestInterface {

        @Nullable
        public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            return inflater.inflate(R.layout.at_test, container, false);
        }

        public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            this.getChildFragmentManager().beginTransaction()
                    .replace(R.id.testAt_frame, new TestImplSupportFragment2())
                    .commit();
        }

        @NonNull
        public final Fragment getChildFragment() {
            return Premisex.requireNotNull(this.getChildFragmentManager().findFragmentById(R.id.testAt_frame));
        }
    }

    public static final class TestImplSupportFragment2 extends Fragment {

    }

    public static class TestFindUserVisibleChildActivity extends FragmentActivity {

        @Override
        protected void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            setContentView(R.layout.at_view_pager);

            ViewPager viewPager = findViewById(R.id.viewPagerAt_viewPager);
            viewPager.setAdapter(new FragmentPagerAdapter(getSupportFragmentManager()) {
                @Override
                public Fragment getItem(int p0) {
                    if (p0 == 2) {
                        return Fragmentx.instantiate(TestFindUserVisibleChildFragment.class, new BundleBuilder().putString("position", String.valueOf(p0)).build());
                    } else {
                        return Fragmentx.instantiate(TestFindUserVisibleChildFragment2.class, new BundleBuilder().putString("position", String.valueOf(p0)).build());
                    }
                }

                @Override
                public int getCount() {
                    return 5;
                }
            });
            viewPager.setCurrentItem(2);
        }
    }

    public static class TestFindUserVisibleChildFragment extends Fragment {
        @Override
        public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.at_view_pager, container, false);
        }

        @Override
        public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);

            ViewPager viewPager = view.findViewById(R.id.viewPagerAt_viewPager);
            viewPager.setAdapter(new FragmentPagerAdapter(getChildFragmentManager()) {
                @Override
                public Fragment getItem(int p0) {
                    return Fragmentx.instantiate(TestFindUserVisibleChildFragment2.class, new BundleBuilder().putString("position", String.valueOf(p0)).build());
                }

                @Override
                public int getCount() {
                    return 5;
                }
            });
            viewPager.setCurrentItem(3);
        }
    }

    public static class TestFindUserVisibleChildFragment2 extends Fragment {
        @Override
        public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return new TextView(getContext());
        }

        @Override
        public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);

            ((TextView) view).setText("position: " + Argsx.readStringArg(this, "position"));
        }
    }
}

File: androidx-kt/src/androidTest/java/me/panpf/androidxkt/test/app/FragmentxTest.kt
/*
 * Copyright (C) 2018 Peng fei Pan <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("DEPRECATION")

package me.panpf.androidxkt.test.app

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.test.rule.ActivityTestRule
import androidx.test.runner.AndroidJUnit4
import me.panpf.androidx.Androidx
import me.panpf.androidx.app.Argsx
import me.panpf.androidx.os.BundleBuilder
import me.panpf.androidxkt.app.*
import me.panpf.androidxkt.test.R
import me.panpf.androidxkt.waitRunInUI
import me.panpf.androidxkt.waitRunInUIResult
import me.panpf.javaxkt.util.requireNotNull
import org.junit.Assert
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class FragmentxTest {

    @get:Rule
    val activityTestRule = ActivityTestRule(TestActivity::class.java)

    @get:Rule
    val findUserVisibleChildActivityRule = ActivityTestRule(TestFindUserVisibleChildActivity::class.java)

    @Test
    fun testGetApplication() {
        val originFragment = activityTestRule.activity.getOriginFragment()
        val supportFragment = activityTestRule.activity.getSupportFragment()

        Assert.assertNotNull(originFragment.getApplication())
        Assert.assertNotNull(supportFragment.getApplication())

        Assert.assertNull(android.app.Fragment().getApplication())
        Assert.assertNull(Fragment().getApplication())

        try {
            android.app.Fragment().requireApplication()
            Assert.fail()
        } catch (ignored: Exception) {
        }

        try {
            Fragment().requireApplication()
            Assert.fail()
        } catch (ignored: Exception) {
        }

    }

    @Test
    fun testDestroyed() {
        val originFragment = activityTestRule.activity.getOriginFragment()
        val supportFragment = activityTestRule.activity.getSupportFragment()

        Assert.assertFalse(originFragment.isDestroyedCompat())
        Assert.assertFalse(supportFragment.isDestroyedCompat())

        activityTestRule.finishActivity()
        Thread.sleep(2000)

        Assert.assertTrue(originFragment.isDestroyedCompat())
        Assert.assertTrue(supportFragment.isDestroyedCompat())
    }

    @Test
    fun testGetImplWithParent() {
        val originFragment = activityTestRule.activity.getOriginFragment() as TestImplOriginFragment
        val supportFragment = activityTestRule.activity.getSupportFragment() as TestImplSupportFragment

        Assert.assertEquals(originFragment.getImplFromParent(ImplTestInterface::class.java).requireNotNull()::class.java, TestImplOriginFragment::class.java)
        Assert.assertEquals(supportFragment.getImplFromParent(ImplTestInterface::class.java).requireNotNull()::class.java, TestImplSupportFragment::class.java)

        if (Androidx.isAtLeast17()) {
            Assert.assertEquals(originFragment.getChildFragment().getImplFromParent(ImplTestInterface::class.java).requireNotNull()::class.java, TestImplOriginFragment::class.java)
        }
        Assert.assertEquals(supportFragment.getChildFragment().getImplFromParent(ImplTestInterface::class.java).requireNotNull()::class.java, TestImplSupportFragment::class.java)

        waitRunInUI {
            activityTestRule.activity.convertChildFragment()
        }

        val originFragment2 = waitRunInUIResult { activityTestRule.activity.getOriginFragment() as TestImplOriginFragment2 }
        val supportFragment2 = waitRunInUIResult { activityTestRule.activity.getSupportFragment() as TestImplSupportFragment2 }

        Assert.assertEquals(originFragment2.getImplFromParent(ImplTestInterface::class.java).requireNotNull()::class.java, TestActivity::class.java)
        Assert.assertEquals(supportFragment2.getImplFromParent(ImplTestInterface::class.java).requireNotNull()::class.java, TestActivity::class.java)

        Assert.assertNull(TestImplOriginFragment2().getImplFromParent(ImplTestInterface::class.java))
        Assert.assertNull(TestImplSupportFragment2().getImplFromParent(ImplTestInterface::class.java))
    }

    @Test
    fun testInstantiate() {
        val originFragment = android.app.Fragment::class.java.instantiateOrigin(Bundle().apply {
            putString("key", "testInstantiate")
        })
        Assert.assertEquals(android.app.Fragment::class.java.name, originFragment::class.java.name)
        Assert.assertEquals("testInstantiate", Argsx.readStringArgOrThrow(originFragment, "key"))

        val supportFragment = androidx.fragment.app.Fragment::class.java.instantiate(Bundle().apply {
            putString("key", "testInstantiate")
        })
        Assert.assertEquals(androidx.fragment.app.Fragment::class.java.name, supportFragment::class.java.name)
        Assert.assertEquals("testInstantiate", Argsx.readStringArgOrThrow(supportFragment, "key"))

        val originFragment2 = android.app.Fragment::class.java.instantiateOrigin()
        Assert.assertEquals(android.app.Fragment::class.java.name, originFragment2::class.java.name)

        val supportFragment2 = androidx.fragment.app.Fragment::class.java.instantiate()
        Assert.assertEquals(androidx.fragment.app.Fragment::class.java.name, supportFragment2::class.java.name)
    }

    @Test
    fun testFindUserVisibleChildFragment() {
        val activity = findUserVisibleChildActivityRule.activity
        // 定义多少个 ActivityTestRule 测试方法执行的时候就会启动多少个 ActivityTestRule 为了让 findUserVisibleChildActivityRule 处于 resumed 状态
        activityTestRule.finishActivity()
        // 此测试要求手机处于屏幕解锁状态

        val fragmentFromActivity = activity.findUserVisibleChildFragment().requireNotNull()
        Assert.assertEquals(TestFindUserVisibleChildFragment::class.java.name, fragmentFromActivity::class.java.name)

        val fragmentFromList = activity.supportFragmentManager.fragments.findUserVisibleChildFragment().requireNotNull()
        Assert.assertEquals(TestFindUserVisibleChildFragment::class.java.name, fragmentFromList::class.java.name)

        val fragmentFromActivity2 = fragmentFromActivity as TestFindUserVisibleChildFragment
        val fragmentFromChildFragment = fragmentFromActivity2.findUserVisibleChildFragment().requireNotNull()
        Assert.assertTrue(fragmentFromChildFragment.tag, fragmentFromChildFragment.tag.orEmpty().startsWith("android:switcher") && fragmentFromChildFragment.tag.orEmpty().endsWith(":3"))
    }

    @Test
    fun testFindFragmentByViewPagerCurrentItem() {
        val activity = findUserVisibleChildActivityRule.activity
        // 定义多少个 ActivityTestRule 测试方法执行的时候就会启动多少个 ActivityTestRule 为了让 findUserVisibleChildActivityRule 处于 resumed 状态
        activityTestRule.finishActivity()
        // 此测试要求手机处于屏幕解锁状态

        val fragmentFromActivity = activity.findFragmentByViewPagerCurrentItem(2).requireNotNull()
        Assert.assertEquals(TestFindUserVisibleChildFragment::class.java.name, fragmentFromActivity::class.java.name)
        Assert.assertTrue(fragmentFromActivity.tag, fragmentFromActivity.tag.orEmpty().startsWith("android:switcher") && fragmentFromActivity.tag.orEmpty().endsWith(":2"))

        val fragmentFromList = activity.supportFragmentManager.fragments.findFragmentByViewPagerCurrentItem(2).requireNotNull()
        Assert.assertEquals(TestFindUserVisibleChildFragment::class.java.name, fragmentFromList::class.java.name)
        Assert.assertTrue(fragmentFromList.tag, fragmentFromList.tag.orEmpty().startsWith("android:switcher") && fragmentFromList.tag.orEmpty().endsWith(":2"))

        val fragmentFromActivity2 = fragmentFromActivity as TestFindUserVisibleChildFragment
        val fragmentFromChildFragment = fragmentFromActivity2.findFragmentByViewPagerCurrentItem(3).requireNotNull()
        Assert.assertTrue(fragmentFromChildFragment.tag, fragmentFromChildFragment.tag.orEmpty().startsWith("android:switcher") && fragmentFromChildFragment.tag.orEmpty().endsWith(":3"))
    }

    class TestActivity : androidx.fragment.app.FragmentActivity(), ImplTestInterface {

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            setContentView(R.layout.at_multi_frame)

            fragmentManager.beginTransaction()
                    .replace(R.id.multiFrameAt_frame1, TestImplOriginFragment())
                    .commit()
            supportFragmentManager.beginTransaction()
                    .replace(R.id.multiFrameAt_frame2, TestImplSupportFragment())
                    .commit()
        }

        fun getOriginFragment(): android.app.Fragment =
                fragmentManager.findFragmentById(R.id.multiFrameAt_frame1).requireNotNull()

        fun getSupportFragment(): androidx.fragment.app.Fragment =
                supportFragmentManager.findFragmentById(R.id.multiFrameAt_frame2).requireNotNull()

        fun convertChildFragment() {
            fragmentManager.beginTransaction()
                    .replace(R.id.multiFrameAt_frame1, TestImplOriginFragment2())
                    .commit()
            supportFragmentManager.beginTransaction()
                    .replace(R.id.multiFrameAt_frame2, TestImplSupportFragment2())
                    .commit()
        }
    }

    interface ImplTestInterface

    class TestImplOriginFragment : android.app.Fragment(), ImplTestInterface {
        override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater?.inflate(R.layout.at_test, container, false)
        }

        override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            if (Androidx.isAtLeast17()) {
                childFragmentManager.beginTransaction()
                        .replace(R.id.testAt_frame, TestImplOriginFragment2())
                        .commit()
            }
        }

        fun getChildFragment(): android.app.Fragment =
                childFragmentManager.findFragmentById(R.id.testAt_frame).requireNotNull()
    }

    class TestImplOriginFragment2 : android.app.Fragment()

    class TestImplSupportFragment : androidx.fragment.app.Fragment(), ImplTestInterface {
        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater.inflate(R.layout.at_test, container, false)
        }

        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            childFragmentManager.beginTransaction()
                    .replace(R.id.testAt_frame, TestImplSupportFragment2())
                    .commit()
        }

        fun getChildFragment(): androidx.fragment.app.Fragment =
                childFragmentManager.findFragmentById(R.id.testAt_frame).requireNotNull()
    }

    class TestImplSupportFragment2 : androidx.fragment.app.Fragment()

    class TestFindUserVisibleChildActivity : androidx.fragment.app.FragmentActivity() {
        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            setContentView(R.layout.at_view_pager)

            val viewPager = findViewById<androidx.viewpager.widget.ViewPager>(R.id.viewPagerAt_viewPager)
            viewPager.adapter = object : androidx.fragment.app.FragmentPagerAdapter(supportFragmentManager) {
                override fun getItem(p0: Int): androidx.fragment.app.Fragment {
                    return if (p0 == 2) {
                        TestFindUserVisibleChildFragment::class.java.instantiate(BundleBuilder().putString("position", p0.toString()).build())
                    } else {
                        TestFindUserVisibleChildFragment2::class.java.instantiate(BundleBuilder().putString("position", p0.toString()).build())
                    }
                }

                override fun getCount(): Int = 5
            }
            viewPager.currentItem = 2
        }
    }

    class TestFindUserVisibleChildFragment : androidx.fragment.app.Fragment() {
        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater.inflate(R.layout.at_view_pager, container, false)
        }

        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)

            val viewPager = view.findViewById<androidx.viewpager.widget.ViewPager>(R.id.viewPagerAt_viewPager)
            viewPager.adapter = object : androidx.fragment.app.FragmentPagerAdapter(childFragmentManager) {
                override fun getItem(p0: Int): androidx.fragment.app.Fragment =
                        TestFindUserVisibleChildFragment2::class.java.instantiate(BundleBuilder().putString("position", p0.toString()).build())

                override fun getCount(): Int = 5
            }
            viewPager.currentItem = 3
        }
    }

    class TestFindUserVisibleChildFragment2 : androidx.fragment.app.Fragment() {
        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return TextView(context)
        }

        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)

            (view as TextView).text = "position: ${readStringArgOrThrow("position")}"
        }
    }
}
